### PR TITLE
Add AIM examples and docs

### LINKED ISSUE #1497: Add AIM examples and docs
An implementation for the AIM model (Scalable Pre-training of Large Autoregressive Image Models) was added in #1479. This issue is about adding examples and docs for the model.

**Todo**
- [ ] Add examples
- [ ] Add docs

### PR DESCRIPTION
**Pull request by guarin**
### Changes

* Add AIM examples
* Add AIM docs
* Add AIM to README

[Uploading AIM — lightly 1.5.0 documentation.pdf…]()

Closes #1497

### CODE REVIEW

File: examples/pytorch/aim.py [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+        batch_size = images.shape[0]
+
+        mask = utils.random_prefix_mask(
+            size=(batch_size, self.num_patches),
+            max_prefix_length=self.num_patches - 1,
+            device=images.device,
+        )
+        features = self.backbone.forward_features(images, mask=mask)
+        # Add positional embedding before head.
+        features = self.backbone._pos_embed(features)
```
Comment (ersi-lightly) [inline examples/pytorch/aim.py:45]: It would be great if we could avoid calling here the private method.. I suppose it is the best we can do though right now, as we need to add again the positional embedding after `forward_features`

File: examples/pytorch/aim.py [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+        predictions = self.projection_head(features)
+
+        # Convert images to patches and normalize them.
+        patches = utils.patchify(images, self.patch_size)
+        mean = patches.mean(dim=-1, keepdim=True)
+        var = patches.var(dim=-1, keepdim=True)
+        patches = (patches - mean) / (var + 1.0e-6) ** 0.5
```
Comment (ersi-lightly) [inline examples/pytorch/aim.py:52]: Did you find that there were numerical instabilities here? Is the addition of 1.0e-6 necessary here?
Also, you could probably use `torch.nn.functional.batch_norm`.

File: examples/pytorch/aim.py [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+
+
+vit = MaskedCausalVisionTransformer(
+    img_size=224,
+    patch_size=32,
+    embed_dim=768,
+    depth=12,
+    num_heads=12,
+    qk_norm=False,
+    class_token=False,
```
Comment (ersi-lightly) [inline examples/pytorch/aim.py:64]: This will set the `vit.cls_token` that I mentioned before

File: examples/pytorch/aim.py [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+dataloader = torch.utils.data.DataLoader(
+    dataset,
+    batch_size=256,
+    shuffle=True,
+    drop_last=True,
+    num_workers=8,
+)
+
+criterion = nn.MSELoss()
+optimizer = torch.optim.AdamW(model.parameters(), lr=1.5e-4)
```
Comment (ersi-lightly) [inline examples/pytorch/aim.py:92]: I would add the learning rate as a variable in the example.

File: examples/pytorch_lightning/aim.py [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+from lightly.transforms import AIMTransform
+
+
+class AIM(pl.LightningModule):
+    def __init__(self) -> None:
+        super().__init__()
+
+        img_size = 224
+        self.patch_size = 32
+        self.num_patches = (img_size // self.patch_size) ** 2
```
Comment (ersi-lightly) [inline examples/pytorch_lightning/aim.py:21]: You can define the `self.num_patches`, after creating an object `vit` of `MaskedCausalVisionTransformer` below using `self.num_patches = vit.patch_embed.num_patches`

Comment (ersi-lightly) [inline examples/pytorch_lightning/aim.py:21]: Similar for `patch_size` and you can use `vit.patch_embed.patch_size[0]`. This way you wouldn't need to have the attribute `patch_size` for the class `AIM`

File: examples/pytorch_lightning/aim.py [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+
+        vit = MaskedCausalVisionTransformer(
+            img_size=img_size,
+            patch_size=self.patch_size,
+            embed_dim=768,
+            depth=12,
+            num_heads=12,
+            qk_norm=False,
+            class_token=False,
+            no_embed_class=True,
```
Comment (ersi-lightly) [inline examples/pytorch_lightning/aim.py:31]: It would be better to pass the values of these parameters at the `init` of `AIM`

File: examples/pytorch_lightning/aim.py [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+        patches = utils.patchify(images, self.patch_size)
+        mean = patches.mean(dim=-1, keepdim=True)
+        var = patches.var(dim=-1, keepdim=True)
+        patches = (patches - mean) / (var + 1.0e-6) ** 0.5
+
+        loss = self.criterion(predictions, patches)
+        return loss
+
+    def configure_optimizers(self):
+        optim = torch.optim.AdamW(self.parameters(), lr=1.5e-4)
```
Comment (ersi-lightly) [inline examples/pytorch_lightning/aim.py:75]: Similar to before, I would pass the learning rate as a parameter to `configure_optimizers`. We could also pass the optimizer and only by default use Adam (but this is not necessary for the example TBH)

File: examples/pytorch_lightning/aim.py [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+        return optim
+
+
+model = AIM()
```
Comment (ersi-lightly) [inline examples/pytorch_lightning/aim.py:79]: Here I would pass the values of the parameters of the model as I mentioned before

File: examples/pytorch_lightning_distributed/aim.py [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+from lightly.transforms import AIMTransform
+
+
+class AIM(pl.LightningModule):
+    def __init__(self) -> None:
+        super().__init__()
+
+        img_size = 224
+        self.patch_size = 32
+        self.num_patches = (img_size // self.patch_size) ** 2
```
Comment (ersi-lightly) [inline examples/pytorch_lightning_distributed/aim.py:21]: Similar to the pytorch lightning example

File: examples/pytorch_lightning_distributed/aim.py [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+            qk_norm=False,
+            class_token=False,
+            no_embed_class=True,
+        )
+
+        # Use absolute positional embedding.
+        pos_embed = utils.get_2d_sincos_pos_embed(
+            embed_dim=vit.embed_dim,
+            grid_size=int(self.num_patches**0.5),
+            cls_token=False,
```
Comment (ersi-lightly) [inline examples/pytorch_lightning_distributed/aim.py:38]: Similar to before

File: examples/pytorch_lightning/aim.py [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+        self.backbone = vit
+        self.projection_head = AIMPredictionHead(
+            input_dim=vit.embed_dim, output_dim=3 * self.patch_size**2, num_blocks=1
+        )
+
+        self.criterion = nn.MSELoss()
+
+    def training_step(self, batch, batch_idx):
+        images, targets = batch[0], batch[1]
+        images = images[0]  # images is a list containing only one view
```
Comment (ersi-lightly) [inline examples/pytorch_lightning/aim.py:52]: Could you rename to reflect that?

File: examples/pytorch_lightning_distributed/aim.py [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+        features = self.backbone.forward_features(images, mask=mask)
+        # Add positional embedding before head.
+        features = self.backbone._pos_embed(features)
+        predictions = self.projection_head(features)
+
+        # Convert images to patches and normalize them.
+        patches = utils.patchify(images, self.patch_size)
+        mean = patches.mean(dim=-1, keepdim=True)
+        var = patches.var(dim=-1, keepdim=True)
+        patches = (patches - mean) / (var + 1.0e-6) ** 0.5
```
Comment (ersi-lightly) [inline examples/pytorch_lightning_distributed/aim.py:69]: similar to before

File: examples/pytorch_lightning_distributed/aim.py [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+        self.backbone = vit
+        self.projection_head = AIMPredictionHead(
+            input_dim=vit.embed_dim, output_dim=3 * self.patch_size**2, num_blocks=1
+        )
+
+        self.criterion = nn.MSELoss()
+
+    def training_step(self, batch, batch_idx):
+        images, targets = batch[0], batch[1]
+        images = images[0]  # images is a list containing only one view
```
Comment (ersi-lightly) [inline examples/pytorch_lightning_distributed/aim.py:52]: similar to before

File: examples/pytorch_lightning_distributed/aim.py [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+        patches = utils.patchify(images, self.patch_size)
+        mean = patches.mean(dim=-1, keepdim=True)
+        var = patches.var(dim=-1, keepdim=True)
+        patches = (patches - mean) / (var + 1.0e-6) ** 0.5
+
+        loss = self.criterion(predictions, patches)
+        return loss
+
+    def configure_optimizers(self):
+        optim = torch.optim.AdamW(self.parameters(), lr=1.5e-4)
```
Comment (ersi-lightly) [inline examples/pytorch_lightning_distributed/aim.py:75]: similar to before

File: examples/pytorch_lightning_distributed/aim.py [hunk 15]
```diff
@@ -0,0 +1,111 @@
...
+        return optim
+
+
+model = AIM()
```
Comment (ersi-lightly) [inline examples/pytorch_lightning_distributed/aim.py:79]: similar to before

**Review by ersi-lightly — CHANGES_REQUESTED** [review]

File: examples/pytorch/aim.py [hunk 16]
```diff
@@ -0,0 +1,107 @@
...
+    def __init__(self, vit):
+        super().__init__()
+
+        self.patch_size = vit.patch_embed.patch_size[0]
+        self.num_patches = vit.patch_embed.num_patches
+
+        # Use absolute positional embedding.
+        pos_embed = utils.get_2d_sincos_pos_embed(
+            embed_dim=vit.embed_dim,
+            grid_size=int(self.num_patches**0.5),
```
[the remark this answers is not in the document]
Comment (guarin) [reply examples/pytorch/aim.py:24]: I changed it to use your `utils.initialize_2d_sine_cosine_positional_embedding` function instead. Much better!

File: examples/pytorch/aim.py (hunk 2 above)
Comment (guarin) [reply examples/pytorch/aim.py:52]: No this is from the MAE codebase: https://github.com/facebookresearch/mae/blob/efb2a8062c206524e35e47d04501ed4f544c0ae8/models_mae.py#L206-L209

I added `utils.normalize_mean_var` and used this instead. Note that we can't use `torch.nn.functional.batch_norm` because it expects more inputs, see: https://pytorch.org/docs/stable/generated/torch.nn.functional.batch_norm.html

File: examples/pytorch/aim.py (hunk 1 above)
Comment (guarin) [reply examples/pytorch/aim.py:45]: Yes, maybe we can change it in the future to use something like `add_pos_embed` from the masked vision transformer.

File: examples/pytorch/aim.py (hunk 4 above)
Comment (guarin) [reply examples/pytorch/aim.py:92]: I would leave it as is, we have it always hardcoded in the examples.

File: examples/pytorch_lightning/aim.py (hunk 5 above)
Comment (guarin) [reply examples/pytorch_lightning/aim.py:21]: Good point!

File: examples/pytorch_lightning/aim.py (hunk 6 above)
Comment (guarin) [reply examples/pytorch_lightning/aim.py:31]: I will keep them hardcoded. The examples are optimized to be short and concise, we do the same for all the other examples.

File: examples/pytorch/aim.py (hunk 4 above)
Comment (ersi-lightly) [reply examples/pytorch/aim.py:83]: hahaha I feel like we keep having the same conversation everytime!

File: lightly/models/utils.py [hunk 17]
```diff
@@ -638,13 +639,15 @@ def add_stochastic_depth_to_blocks(vit: Module, prob: float = 0.0, mode="row") -
             mod.mlp = Sequential(mod.mlp, StochasticDepth(p=prob, mode=mode))
 
 
-def initialize_2d_sine_cosine_positional_embedding(pos_embedding: Parameter) -> None:
+def initialize_2d_sine_cosine_positional_embedding(
+    pos_embedding: Parameter, has_class_token: bool
```
Comment (ersi-lightly) [inline lightly/models/utils.py:643]: Could you make `has_class_token` optional? In the other parts of the code where we use this we do not pass a value for `has_class_token`

Comment (guarin) [reply lightly/models/utils.py:643]: I updated the other parts of the code. I think it is better to not have it optional because it is easy to miss.

File: examples/pytorch/aim.py (hunk 2 above)
Comment (ersi-lightly) [reply examples/pytorch/aim.py]: great, thanks!

File: lightly/models/utils.py (hunk 17 above)
Comment (ersi-lightly) [reply lightly/models/utils.py:643]: ok! If you are sure that you have updated all affected parts, feel free to resolve.

**Review by ersi-lightly — APPROVED** [review]
